app/src/main/kotlin/com/wire/android/util/FileUtil.kt

```diff
@@ -39,16 +48,49 @@ suspend fun Uri.toByteArray(context: Context): ByteArray {
     }
 }
 
-fun getWritableImageAttachment(context: Context) = getTempWritableAttachmentUri(context, TEMP_IMG_ATTACHMENT_FILENAME)
-
-fun getWritableVideoAttachment(context: Context) = getTempWritableAttachmentUri(context, TEMP_VIDEO_ATTACHMENT_FILENAME)
+fun Context.getTempWritableImageUri() = getTempWritableAttachmentUri(this, TEMP_IMG_ATTACHMENT_FILENAME)
+fun Context.getTempWritableVideoUri() = getTempWritableAttachmentUri(this, TEMP_VIDEO_ATTACHMENT_FILENAME)
 
 private fun getTempWritableAttachmentUri(context: Context, fileName: String): Uri {
     val file = File(context.cacheDir, fileName)
     file.setWritable(true)
     return FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".provider", file)
 }
 
+private fun Context.saveFileDataToDownloadsFolder(downloadedFile: File, fileSize: Int): Uri? {
+    val resolver = contentResolver
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        val contentValues = ContentValues().apply {
+            put(MediaStore.MediaColumns.DISPLAY_NAME, downloadedFile.name)
+            put(MediaStore.MediaColumns.MIME_TYPE, Uri.parse(downloadedFile.path).getMimeType(this@saveFileDataToDownloadsFolder))
+            put(MediaStore.MediaColumns.SIZE, fileSize)
+        }
+        resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, contentValues)
+    } else {
+        val authority = "${packageName}.provider"
+        val destinyFile = File(getExternalFilesDir(DIRECTORY_DOWNLOADS), downloadedFile.name)
+        FileProvider.getUriForFile(this, authority, destinyFile)
+    }?.also { downloadedUri ->
+        resolver.openOutputStream(downloadedUri).use { outputStream ->
+            val brr = ByteArray(DATA_COPY_BUFFER_SIZE)
+            var len: Int
+            val bufferedInputStream = BufferedInputStream(FileInputStream(downloadedFile.absoluteFile))
+            while ((bufferedInputStream.read(brr, 0, brr.size).also { len = it }) != -1) {
+                outputStream?.write(brr, 0, len)
+            }
+            outputStream?.flush()
+            bufferedInputStream.close()
+        }
+    }
+}
+
+fun Context.copyDataToTempFile(assetName: String, assetData: ByteArray): Uri {
+    val file = File(cacheDir, assetName)
+    file.setWritable(true)
+    file.writeBytes(assetData)
+    return FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID + ".provider", file)
```

Comment:
We have this `BuildConfig.Application_ID + ".provider"` in at least 3 places (other than the Manifest, of course).

Can we create a constant somewhere, document it (linking to Manifest) in order to reduce repetition and make it easier to understand where it comes from?

Reply:
Great catch! I just refactored 5 of them into a common one 🙈